src/react-demo/Content/src/Hmr.fs

```diff
@@ -0,0 +1,71 @@
+module Hmr
+
+open Elmish
+open Elmish.Browser.Navigation
+open Elmish.Browser.UrlParser
+open Fable.Core
+open Fable.Core.JsInterop
+open Fable.Import
+open Fable.Import.Browser
+open App.State
+open Fable.Helpers.React
+open Fable.Helpers.React.Props
+
+module Program =
+
+    type IModule =
+        abstract hot: obj with get, set
+
+    let [<Global>] [<Emit("module")>] ``module`` : IModule = jsNative
+
+    type HMRMsg<'msg> =
+        | UserMsg of 'msg
+        | Reload
+
+    type HMRModel<'model> =
+        { HMRCount : int
+          UserModel : 'model }
+
+    let mutable hmrState : obj = null
+
+    let inline withHMR (program:Elmish.Program<'arg, 'model, 'msg, 'view>) =
+
+        if not (isNull ``module``.hot) then
+            ``module``.hot?accept() |> ignore
+
+        let map (model, cmd) =
+            model, cmd |> Cmd.map UserMsg
+
+        let update msg model =
+            let newModel,cmd =
+                match msg with
+                | UserMsg msg ->
+                    let newModel, cmd = program.update msg model.UserModel
+                    { model with UserModel = newModel }, cmd
+                | Reload ->
+                    { model with HMRCount = model.HMRCount + 1 }, Cmd.none
+                |> map
+
+            hmrState <- newModel
+            // Store the state
```

Comment:
where doees it store the state?

Reply:
It's stored inside a variable of `Hmr.fs` module. So in memory. This should works if you don't have several Elmish application running inside your page.

The other solution, I have in mind is something like:

```fs
Program.mkProgram ....
|> Program.withHmr "myApplicationState"
|> Program.run
```

And use `"myApplicationState"` to store the state into the `window` scope. (Ex: `window?("myApplicationState")`)

Reply:
how would that work with RN?

Reply:
For RN, we would need to determine a storage for the application state.

This storage need to exist only during the application runtime. If you reload your app then it's destroyed and so your application start from scratch. It's important, because if you store your application in a persistent (like `LocalStorage`) place then you would need to clean it by hand.

Reply:
@forki Seems like React Native have something called `global` instead of `window`.

https://github.com/facebook/react-native/issues/740

Reply:
so that would be easy change?

Reply:
Yes, or we could even use the polyfill `window = global || window` so no adaptation is required for RN.

Reply:
but where do you use "window" at all?

Reply:
fuck. I just copied your module into my RN app. and it looks like it works ootb. can this be true?

Reply:
In the current implementation we don't use `window`.

We are using a [local module variable](https://github.com/fable-elmish/templates/blob/43ce9f08d4c29f31a93a8d4913526b08bcab7c71/src/react-demo/Content/src/Hmr.fs#L22). And I think, we will keep it like that.

Just for @et1975 I was speaking about using customisable variable inside window scope for people running several applications at the same time. But as we discussed, we are not supporting this feature for now so using a local variable is enough.

And, as I am probably the only one doing that I can make my own version of the HMR module for my work apps.